flang/lib/Optimizer/Builder/IntrinsicCall.cpp

```diff
@@ -5911,6 +5917,37 @@ IntrinsicLibrary::genReduce(mlir::Type resultType,
   return readAndAddCleanUp(resultMutableBox, resultType, "REDUCE");
 }
 
+// RENAME
+fir::ExtendedValue
+IntrinsicLibrary::genRename(std::optional<mlir::Type> resultType,
+                            mlir::ArrayRef<fir::ExtendedValue> args) {
+  assert((args.size() == 3 && !resultType.has_value()) ||
+         (args.size() == 2 && resultType.has_value()));
+
+  mlir::Value path1 = fir::getBase(args[0]);
+  mlir::Value path2 = fir::getBase(args[1]);
+  if (!path1 || !path2)
+    fir::emitFatalError(loc, "Expected at least two dummy arguments");
+
+  if (resultType.has_value()) {
+    // code-gen for the function form of RENAME
+    auto statusAddr = builder.createTemporary(loc, *resultType);
+    auto statusBox = builder.createBox(loc, statusAddr);
+    fir::runtime::genRename(builder, loc, path1, path2, statusBox);
+    return builder.create<fir::LoadOp>(loc, statusAddr);
+  } else {
+    // code-gen for the procedure form of RENAME
+    mlir::Type boxNoneTy = fir::BoxType::get(builder.getNoneType());
+    auto status = args[2];
+    mlir::Value statusBox =
+        isStaticallyPresent(status)
+            ? fir::getBase(status)
+            : builder.create<fir::AbsentOp>(loc, boxNoneTy).getResult();
```

Comment:
Earlier you defined the intrinsic as supporting a dynamically optional argument. To support this you will also need a runtime check for something like this

```
subroutine doRename(status)
  integer, optional :: status
  call rename("old", "new", status)
end subroutine
```

Here we cannot determine statically whether `status` is present, but if it is present it should be used. I think the current code would treat it as absent. For example, see the implementation of C_ASSOCIATED in this file.

Reply:
I have this example now:

```Fortran
module ren                                                                         
    contains                                                                       
        subroutine rename_opt(src, dst, status)                                    
            implicit none                                                          
            character(*) :: src                                                    
            character(*) :: dst                                                    
            integer, optional :: status                                            
                                                                                   
            call rename(src, dst, status)                                          
        end subroutine rename_opt                                                  
end module ren                                                                     
```

The `isStaticallyPresent` still says that `status` is present (which I what I would expect, too, as the argument is there in the code).  So, it seems like the handling of `OPTIONAL` seems to be done at the call site.  Would be an accurate assumption to make for the code-gen in this PR?

Reply:
Here's a bit of MLIR that shows what's going on:

```
    %5:2 = hlfir.declare %arg2 dummy_scope %0 {fortran_attrs = #fir.var_attrs<optional>, uniq_name = "_QMrenFren>
    %6 = fir.is_present %5#0 : (!fir.ref<i32>) -> i1
    %9 = fir.embox %5#1 : (!fir.ref<i32>) -> !fir.box<i32>
    %10 = fir.absent !fir.box<i32>
    %11 = arith.select %6, %9, %10 : !fir.box<i32>
```

This looks like the code-gen for `OPTIONAL` is already doing something sensible, right?

Reply:
LGTM, sorry for the confusion